### PR TITLE
rework how hooks in `error-stack` emit lines and snippets.

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

In a conversation with @TimDiekmann we decided to move away from the current approach of `vec![Emit]` being returned, to using methods on `HookContext` instead. 

The initial reason mainly was to couple snippets and emit and make them either defer or emit immediately together, otherwise, the ordering of snippets would be confusing.

## 🔗 Related links

- Discussion: https://github.com/hashintel/hash/pull/974#discussion_r957172889

## 🔍 What does this change?

The current proposal is: 

- [x] `Emit` moves to an implementation detail
- [x] rename `Emit::Next` to `Emit::Immediate` or `Emit::Yield`
- [x] Use methods on HookContext instead
- [x] remove the reversal of `Emit::Defer` attachments

## 📜 Does this require a change to the docs?

All examples that currently use the old function signature and explanation of the usage need to be adjusted, as well as the examples.

#### Proposal 1: "Two different worlds"

```rust
install_hook(|var, context| -> () {
    context.emit(Suggestion(var));
    context.emit_deferred("do better next time :upside_down_face: ");

    context.snippet("My snippet goes here");
	context.snippet_deferred("My deferred snippet goes here");
})
```

- 👍 easy to implement
- 👍 short names
- 😐 does not enforce that snippets need to be paired with emit
- 👎 confusing for users that `snippet_deferred` and `emit_deferred` need to be used together
- 👎 One can easily `snippet` and `emit_deffered`, which is unusual and shouldn't happen/be allowed IMO

#### Proposal 2: "Long and explicit"

```rust
install_hook(|var, context| -> () {
    context.emit(Suggestion(var));
    context.emit_deferred("do better next time :upside_down_face: ");

	context.emit_with_snippet(Suggestion(var), "My snippet goes here");
	context.emit_with_snippet_deferred("do better next time :upside_down_face:", "My deferred snippet goes here");
})
```

- 👍 easy to implement
- 👍 enforces that snippets need to be paired with emit
- 👍 one can only defer snippet+emit, but not just emit a snippet
- 👎 long names
- ❓ What about multiple snippets per emit?!

#### Proposal 3: "Builder pattern"

```rust
install_hook(|var, context| -> () {
	context.emit(Suggestion(var)).with_snippet("My snippet goes here").next();
	context.emit("do better next time :upside_down_face:").with_snippet("My deferred snippet goes here").defer();
})

// OR

install_hook(|var, context| -> Vec<Emit> {
	vec![
		context.emit(Suggestion(var)).with_snippet("My snippet goes here").next(),
		context.emit("do better next time :upside_down_face:").with_snippet("My deferred snippet goes here").defer()
	]
})
```

- 👍 enforces that snippets need to be paired with emit
- 👍 enables multiple snippets paired with emit, but cannot emit snippet w/o emit
- 😐 single API for both `defer` and `yield/next` (only finish method is different)
- 👎 a bit harder to implement
- 👎 @TimDiekmann really dislikes the builder pattern^^
- ❓ What to do if one drops and does not finish via `.next()` or `.defer()`

### What to rename `Emit::Next` to?

* `Emit::Yield` (like generators)
* `Emit::Immediate` (long name, but used in computer graphics)

## 🛡 What tests cover this?

The existing tests.

## 📹 Demo

Look at the rendered documentation

### CODE REVIEW

**Comment by TimDiekmann** [comment]
> #### Proposal 1: "Two different worlds"
> [...]
> * 😐 does not enforce that snippets need to be paired with emit

Do we want to require it?

> * 👎 confusing for users that `snippet_deferred` and `emit_deferred` need to be used together
> * 👎 One can easily `snippet` and `emit_deffered`, which is unusual and shouldn't happen/be allowed IMO

Do we want to force the user to do that? Maybe they have good reasons 🤔 

---

> #### Proposal 2: "Long and explicit"
> [...]
> * 👎 long names

💯 

> * ❓ What about multiple snippets per emit?!

This feels very specific and verbose. It's hard to parse this by looking at the code IMO

---

> #### Proposal 3: "Builder pattern"
> [...]
> * 👎 @TimDiekmann really dislikes the builder pattern^^

I don't like the builder pattern in general, but it feels overengineered here. We only have two possibilities: `next` and `deferred`, those cannot be combined.

> * ❓ What to do if one drops and does not finish via `.next()` or `.defer()`

This can be enforced by returning `enum Emit { Immediate, Deferred }` (empty variants). An alternative is to make the intermediate struct `#[must_use]`. As we need intermediate structs anyway we can also use the `enum Emit` approach with empty variants.


I'd probably go with the first or the third approach. The 👎🏻 on the first approach I actually see as an advantage, so you can choose if you want to defer it or not.


The most common use case is probably to only `Emit::next` (or whatever approach we choose), so calling this should be as easy as possible. A stateless approach is probably the easiest way to deal with, for the library and for the user, so I'd go with 1.

---

> ### What to rename `Emit::Next` to?
> * `Emit::Yield` (like generators)
> * `Emit::Immediate` (long name, but used in computer graphics)

`Yield` does not reflect when it's emitted, a deferred message is also yielded, but now immediate 😉

**Comment by indietyp** [comment]
Yea, you are probably right. Lets's go with one. My only fear is that one might be confused that they did `emit_defer` and then `snippet` and are now wondering why it's OoO, but after sleeping over it, I think that's a tradeoff worth having!

**Comment by indietyp** [comment]
@TimDiekmann or @Alfred-Mountfield would you please review this?

(looks larger because it updated snaps)

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 1]
```diff
@@ -829,66 +677,101 @@ impl Opaque {
     }
 }
 
-fn debug_attachments(
-    loc: Option<Line>,
-    position: Position,
+fn debug_attachments_order(emits: impl Iterator<Item = Emit>) -> Vec<String> {
+    let mut out = vec![];
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:681]:
```suggestion
    let mut out = Vec::with_capacity(emits.size_hint().0);
```

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 2]
```diff
@@ -829,66 +677,101 @@ impl Opaque {
...
+            Emit::Defer(value) => {
+                defer.push(value);
+            }
+            Emit::Immediate(value) => {
+                out.push(value);
+            }
+        }
+    }
+
+    out.append(&mut defer);
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:695]: Shouldn't we use
```rust
out.extend(defer);
```
to consume `defer`?
(Maybe I'm missing something here, this is untested)

Comment (indietyp) [reply packages/libs/error-stack/src/fmt/mod.rs:695]: We can, but `.append()` is more efficient.

> The difference between the functions is, that extend adds items one after another while, append adds all of them at once. This allows the destination data structure to decide how to add them efficiently. However, append is a bit less flexible than extend, as you always add the entire second data structure while the iterator of extend can be used to filter items.

from: https://github.com/rust-lang/rust-clippy/issues/4321

Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/mod.rs:751]: I didn't know that `append` does move the elements. I think I never used that function :D

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 3]
```diff
@@ -34,9 +33,9 @@
 //! [`Report::install_debug_hook`].
 //! This type needs to be `'static`, [`Send`], and [`Sync`].
 //!
-//! The hook function must return [`Vec<Emit>`], which decides what is going
-//! to be emitted during printing, refer to the documentation of [`Emit`] for further
-//! information.
+//! You can emit lines, which decide what is going to be emitted during printed with
+//! [`HookContext::emit`] and [`HookContext::emit_deferred`], refer to the documentation of
+//! [`HookContext`] for further information.
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:38]:
```suggestion
//! You can emit lines, which decide what is going to be emitted during printing with
//! [`HookContext::emit`] and [`HookContext::emit_deferred`], refer to the documentation of
//! [`HookContext`] for further information.
```

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 4]
```diff
@@ -54,34 +52,40 @@
...
 //!     }
 //!
-//!     vec![Emit::next(format!("Warning ({idx}) occurred"))]
+//!     ctx.emit(format!("Warning ({idx}) occurred"));
 //!  });
 //!
-//! // you can use arbitrary values as arguments, just make sure that you won't repeat them.
-//! // here we use [`Emit::defer`], this means that this value will be put at the end of the group.
-//! Report::install_debug_hook::<Info>(|Info(val), _| vec![Emit::defer(format!("Info: {val}"))]);
+//! // here we use [`emit_defer()`], this means that this value will be put at the end of the group.
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:85]:
```suggestion
//! // here we use `emit_deferred()`, this means that this value will be put at the end of the group.
```

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 5]
```diff
@@ -166,187 +170,30 @@ use owo_colors::{OwoColorize, Stream::Stdout, Style as OwOStyle};
...
 #[must_use]
 #[non_exhaustive]
-pub enum Emit {
+pub(crate) enum Emit {
     /// Line is going to be emitted after all immediate lines have been emitted from the current
     /// stack.
     /// This means that deferred lines will always be last in a group.
-    #[cfg_attr(not(any(feature = "std", feature = "spantrace")), allow(dead_code))]
+    // Reason: We could in theory dead-code, but that would hurt readability in the rendering
+    // process.
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:181]: A word is missing here

File: packages/libs/error-stack/src/fmt/mod.rs [hunk 6]
```diff
@@ -885,66 +732,101 @@ impl Opaque {
...
+                (vec![Emit::Immediate(attachment.to_string())], vec![])
             }
         })
         .enumerate()
-        .flat_map(|(idx, value)| {
+        .map(|(idx, (emits, snippets))| {
             // increase the opaque counter, if we're unable to determine the actual value of the
             // frame
-            if idx > 0 && value.is_empty() {
+            if idx > 0 && emits.is_empty() {
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/mod.rs:782]: I think it's a matter of taste, but do we also check, if `snippets.is_empty()` as well? Having a snippet at least means that the hook was called.

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 7]
```diff
@@ -37,59 +39,131 @@ impl HookContextImpl {
...
-/// ## Emitting Snippets
+/// ## Emitting Snippets and Lines
 ///
 /// A [`Debug`] backtrace consists of two different sections, a rendered tree of objects and
 /// additional text/information that is too large to fit into the tree.
 ///
-/// Snippets can be added to the current output via [`attach_snippet()`].
+/// Lines to the rendered tree of objects can be attached via [`HookContext::emit`], or
+/// [`HookContext::emit_deferred`]. Snippets can be added to the current output via
+/// [`HookContext::snippet`], or [`HookContext::snippet_deferred`].
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:98]:
```suggestion
/// Lines can be attached to the rendered tree of objects via [`HookContext::emit`], or
/// [`HookContext::emit_deferred`]. Snippets can be added to the current output via
/// [`HookContext::snippet`], or [`HookContext::snippet_deferred`].
```

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 8]
```diff
@@ -37,59 +39,131 @@ impl HookContextImpl {
...
+}
+
 /// Carrier for contextual information used across hook invocations.
 ///
 /// `HookContext` has two fundamental use-cases:
 /// 1) Emitting Snippets
 /// 2) Storage
 ///
-/// ## Emitting Snippets
+/// ## Emitting Snippets and Lines
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:91]:
```suggestion
/// ## Emitting lines and snippets
```

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 9]
```diff
@@ -37,59 +39,131 @@ impl HookContextImpl {
...
+/// // You can emit a line, which is going to be emitted immediately.
+/// Report::install_debug_hook::<HttpResponseStatusCode>(|HttpResponseStatusCode(val), ctx| {
+///     // This is going to be emitted immediately in the list of snippets.
 ///     if ctx.alternate() {
-///         ctx.attach_snippet(format!("Error {code}:\n  {reason}"));
+///         ctx.snippet(format!("Error {val}: {} Error", if *val < 500 {"Client"} else {"Server"}))
 ///     }
 ///
-///     vec![Emit::next(format!("Error {code}"))]
+///     ctx.emit(format!("Error Code: {val}"));
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:133]:
```suggestion
///     ctx.emit(format!("Error code: {val}"));
```

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 10]
```diff
@@ -524,14 +828,14 @@ pub(crate) struct Hooks {
...
-    pub(crate) fallback: Option<BoxedHook>,
+    pub(crate) fallback: Option<BoxedFallbackHook>,
 }
 
 #[cfg(feature = "std")]
 impl Hooks {
     pub(crate) fn insert<T: Send + Sync + 'static>(
         &mut self,
-        hook: impl for<'a> Fn(&T, &mut HookContext<'a, T>) -> Vec<Emit> + Send + Sync + 'static,
+        hook: impl for<'a> Fn(&T, &mut HookContext<'a, T>) + Send + Sync + 'static,
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:838]: I wonder if `'a` can be inferred? (This applies to other `impl Fn` signatures here as well. I haven't tested it)

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 11]
```diff
@@ -37,59 +39,131 @@ impl HookContextImpl {
...
-///         reason: "Bad Request - Server cannot or will not process request",
-///     });
+/// // You can emit a line which is deferred to the end of the stack.
+/// Report::install_debug_hook::<Suggestion>(|Suggestion(val), ctx| {
+///     let idx = ctx.increment();
+///
+///     // deferred snippets are deferred until the end of snippets of the current stack
+///     if ctx.alternate() {
+///         ctx.snippet_deferred(format!("Suggestion {idx}:\n  {val}"));
+///     }
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:143]: This is not shown in the provided output. I'm fine with that, just want to flag in case you want to change this.

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 12]
```diff
@@ -189,8 +265,237 @@ impl<T> HookContext<'_, T> {
...
+    ///
+    /// struct Error {
+    ///     code: usize,
+    ///     reason: &'static str,
+    /// }
+    ///
+    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
+    ///     if ctx.alternate() {
+    ///         ctx.snippet(format!("Error {code}:\n  {reason}"));
+    ///     }
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:286]: Same as above. We probably only want to show the `.alternate()` once, but I don't really mind either way.

File: packages/libs/error-stack/src/hook.rs [hunk 13]
```diff
@@ -292,7 +294,7 @@ impl Report<()> {
     /// </pre>
     #[cfg(feature = "std")]
     pub fn install_debug_hook_fallback(
-        hook: impl for<'a> Fn(&Frame, &mut HookContext<'a, Frame>) -> Vec<Emit> + Send + Sync + 'static,
+        hook: impl for<'a> Fn(&Frame, &mut HookContext<'a, Frame>) + Send + Sync + 'static,
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/hook.rs:297]: Untested, but the above is working
```suggestion
        hook: impl Fn(&Frame, &mut HookContext<Frame>) + Send + Sync + 'static,
```

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: packages/libs/error-stack/src/fmt/mod.rs (hunk 6 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/mod.rs:782]: Well, this is used for the "opaque" counter. If something returned a snippet, it would not show up in the tree, meaning it is still opaque 🤔

File: packages/libs/error-stack/src/hook.rs (hunk 13 above)
Comment (indietyp) [reply packages/libs/error-stack/src/hook.rs]: fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)

File: packages/libs/error-stack/src/fmt/hook.rs (hunk 12 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:297]: fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)

File: packages/libs/error-stack/src/fmt/hook.rs (hunk 11 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:142]: fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)

File: packages/libs/error-stack/src/fmt/hook.rs (hunk 10 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs]: fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)

File: packages/libs/error-stack/src/fmt/mod.rs (hunk 5 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/mod.rs]: fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 14]
```diff
@@ -37,59 +39,130 @@ impl HookContextImpl {
...
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct HookContextInner<'a> {
+    inner: &'a mut HookContextImpl,
+
+    emits: Vec<Emit>,
+    snippets: Vec<Emit>,
+}
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:54]: Why we need another struct here? Also, when `HookContextImpl` needs to be a. `&mut`? Removing the reference would remove the lifetime on `HookContext` as well and I don't see why we ever need to use `HookContextImpl` at all, this should be an implementation detail to allow casting with`repr(transparent)`.

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 15]
```diff
@@ -37,59 +39,130 @@ impl HookContextImpl {
...
+
+    fn storage_mut(&mut self) -> &mut Storage {
+        &mut self.inner.storage
+    }
+
+    fn alternate(&self) -> bool {
+        self.inner.alternate
+    }
+
+    #[allow(clippy::missing_const_for_fn)]
```
[the remark this answers is not in the document]
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:70]: Yes

[the remark this answers is not in the document]
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:70]: adjusted in [9c774ca](https://github.com/hashintel/hash/pull/1008/commits/9c774ca1fda269d1b2d055e262dd35eaae10ee24)

File: packages/libs/error-stack/src/fmt/hook.rs (hunk 14 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:54]: This is because there are 2 different layers here at play: Debug level invocation and Hook level invocation.

`HookContextInner` is instantiated for every hook, but `HookContextImpl` is only instantiated once for every `Debug` invocation.

This is why `HookContextImpl` hosts `alternate` and `storage`, while `HookContextInner` only has emits and snippets.

The `snippets` property is here to make `return` statements a bit easier in the debug render code, as otherwise, the conditional compilation step would be an utterly unreadable mess, basically we destruct `HookContextInner`, reorder the snippets and emits, then append the snippets to the `HookContextImpl::snippets` and return the emits.

`storage` is the same across the whole `Debug` invocation, not only hook, you can add new data to it, which means it needs to be `mut`.

We could, in theory, move this into `HookContext`, but then wouldn't be able to `repr(transparent)` cast.

Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/hook.rs:54]: I see. I experimented around with this a little bit and moved everything from `HookContextImpl` to `HookContextInner` (so removing the lifetime). I changed the functions to take `HookContext<Frame>` instead of `HookContextImpl`. Instead of `as_hook_context` I reused the `HookContext`, but changed the `into_parts` method to
```rust
    fn take_emits(&mut self) -> (Vec<Emit>, Vec<Emit>) {
        (mem::take(&mut self.emits), mem::take(&mut self.snippets))
    }
```

I opened a PR for this change, so you can see my proposed changes 🙂 
- https://github.com/indietyp/hash/pull/2

File: packages/libs/error-stack/src/fmt/mod.rs (hunk 6 above)
Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/mod.rs:783]: I know what this is for, I was just wondering if we still want to count it as opaque if it emitted a snippet.

File: packages/libs/error-stack/src/fmt/hook.rs (hunk 14 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs]: PR was merged 🎉

File: packages/libs/error-stack/src/fmt/mod.rs (hunk 6 above)
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/mod.rs:780]: I'd prefer to keep it like it is right now, as the opaque attachments counter does not say a hook wasn't called, but rather that it didn't show up. 

(That's at least how it works in my mind)

but that isn't a strong preference, so I'd be happy to change it if you prefer it the other way.

Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/mod.rs:780]: Cc @hashintel/error-stack?

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 16]
```diff
@@ -175,26 +249,271 @@ impl HookContextImpl {
...
+    ///
+    /// struct Error {
+    ///     code: usize,
+    ///     reason: &'static str,
+    /// }
+    ///
+    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
+    ///     if ctx.alternate() {
+    ///         ctx.snippet(format!("Error {code}:\n  {reason}"));
+    ///     }
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:302]: Trying to highlight the actual function in the example
```suggestion
    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
    ///     if ctx.alternate() {
    ///         // Add a snippet to the output
    ///         ctx.snippet(format!("Error {code}:\n  {reason}"));
    ///     }
```

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 17]
```diff
@@ -175,26 +249,271 @@ impl HookContextImpl {
...
+    ///     code: usize,
+    ///     reason: &'static str,
+    /// }
+    ///
+    /// struct Suggestion(&'static str);
+    ///
+    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
+    ///     if ctx.alternate() {
+    ///         ctx.snippet_deferred(format!("Error {code}:\n  {reason}"));
+    ///     }
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:368]:
```suggestion
    ///     if ctx.alternate() {
    ///         // Add a snippet, which is printed after other snippets
    ///         ctx.snippet_deferred(format!("Error {code}:\n  {reason}"));
    ///     }
```

Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/hook.rs:368]: Would be great if you could look for other places as well, I think especially the `cast()` option is a good place to add some explanations as well.

**Review by TimDiekmann — COMMENTED** [review]
Found a few places, where we could slightly improve code quality and documentation. Happy to merge after you re-requested a review 🙂

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 18]
```diff
@@ -175,26 +249,271 @@ impl HookContextImpl {
...
+// TODO: ideally we would want to make `HookContextInner` private, as it is an implementation
+//  detail, but "attribute privacy" as outlined in https://github.com/rust-lang/rust/pull/61969
+//  is currently not implemented for repr(transparent).
 #[repr(transparent)]
-pub struct HookContext<'a, T> {
-    parent: &'a mut HookContextImpl,
-    _marker: PhantomData<T>,
+pub struct HookContext<T> {
+    inner: HookContextInner,
+    _marker: PhantomData<fn(&T)>,
```
[the remark this answers is not in the document]
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/hook.rs:258]: fixed in [81e5df6](https://github.com/hashintel/hash/pull/1008/commits/81e5df68448138db8d6332d62cb43c12b6ba1795)

[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/libs/error-stack/src/fmt/hook.rs:258]: Looking at the code again: This will probably fail and needs to be changed to `fn() -> *const C`, but let's look at the CI 🙃

File: packages/libs/error-stack/src/fmt/hook.rs [hunk 19]
```diff
@@ -49,47 +76,83 @@ impl HookContextImpl {
...
+///     ctx.emit_deferred(format!("Suggestion ({idx})"));
+/// });
+///
+/// // You can emit multiple lines from the same hook.
+/// Report::install_debug_hook::<Warning>(|Warning(val), ctx| {
+///     ctx.emit("Abnormal program execution detected");
+///     ctx.emit(format!("Warning: {val}"));
+/// });
+///
+/// // By not returning anything you are able to hide an attachment
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/hook.rs:143]:
```suggestion
/// // By not adding anything you are able to hide an attachment
```

**Review by TimDiekmann — APPROVED** [review]
